ksqldb-engine/src/main/java/io/confluent/ksql/physical/pull/PullPhysicalPlanBuilder.java

```diff
@@ -0,0 +1,348 @@
+import io.confluent.ksql.metastore.model.DataSource;
+import io.confluent.ksql.metastore.model.DataSource.DataSourceType;
+import io.confluent.ksql.model.WindowType;
+import io.confluent.ksql.name.SourceName;
+import io.confluent.ksql.parser.tree.AllColumns;
+import io.confluent.ksql.parser.tree.Query;
+import io.confluent.ksql.parser.tree.Select;
+import io.confluent.ksql.parser.tree.SingleColumn;
+import io.confluent.ksql.physical.pull.operators.AbstractPhysicalOperator;
+import io.confluent.ksql.physical.pull.operators.DataSourceOperator;
+import io.confluent.ksql.physical.pull.operators.KeyedTableLookupOperator;
+import io.confluent.ksql.physical.pull.operators.KeyedWindowedTableLookupOperator;
+import io.confluent.ksql.physical.pull.operators.ProjectOperator;
+import io.confluent.ksql.physical.pull.operators.SelectOperator;
+import io.confluent.ksql.physical.pull.operators.WhereInfo;
+import io.confluent.ksql.planner.LogicalPlanNode;
+import io.confluent.ksql.planner.plan.DataSourceNode;
+import io.confluent.ksql.planner.plan.FilterNode;
+import io.confluent.ksql.planner.plan.OutputNode;
+import io.confluent.ksql.planner.plan.PlanNode;
+import io.confluent.ksql.planner.plan.ProjectNode;
+import io.confluent.ksql.query.QueryId;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.schema.ksql.LogicalSchema.Builder;
+import io.confluent.ksql.schema.ksql.PhysicalSchema;
+import io.confluent.ksql.schema.ksql.SystemColumns;
+import io.confluent.ksql.schema.ksql.types.SqlType;
+import io.confluent.ksql.schema.ksql.types.SqlTypes;
+import io.confluent.ksql.serde.connect.ConnectSchemas;
+import io.confluent.ksql.services.ServiceContext;
+import io.confluent.ksql.statement.ConfiguredStatement;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.PersistentQueryMetadata;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.kafka.connect.data.ConnectSchema;
+import org.apache.kafka.connect.data.Field;
+import org.apache.kafka.connect.data.Struct;
+
+// CHECKSTYLE_RULES.OFF: ClassDataAbstractionCoupling
+public class PullPhysicalPlanBuilder {
+  // CHECKSTYLE_RULES.ON: ClassDataAbstractionCoupling
+
+  private final MetaStore metaStore;
+  private final KsqlConfig config;
+  private final ServiceContext serviceContext;
+  private final KsqlExecutionContext executionContext;
+  private final Stacker contextStacker;
+  private final ImmutableAnalysis analysis;
+  private final RoutingFilterFactory routingFilterFactory;
+  private final RoutingOptions routingOptions;
+  private final ConfiguredStatement<Query> statement;
+
+  private WhereInfo whereInfo;
+  private PersistentQueryMetadata persistentQueryMetadata;
+  private  QueryId queryId;
+  private Materialization mat;
+  private List<Struct> keys;
+
+  public PullPhysicalPlanBuilder(
```

Comment:
Javadoc for this class regarding the common pattern of the generated physical plan (e.g. should it always be a scan -> filter), what validation it does, what metadata (e.g. materialization) it relies on would be highly appreciated :)

Reply:
The metadata/validation information is going to change with the next PR so I'd rather add it when the code is done. I don't want to end up with wrong javadocs.